File: frontend/src/api/plotAttributeWorkbook.ts
import request from '@/api/request'
import type {
  PlotAttributeImportBatch,
  PlotAttributeImportBatchList,
  PlotAttributeWorkbookExportPayload,
} from '@/types/plotAttributeWorkbook'

/**
 * 导出包含当前版本和逐图斑属性的 XLSX。
 * @param taskCode 作业任务编号。
 * @param payload 操作人和可选显式图斑范围。
 * @returns 可编辑并可回写的 Excel 实体。
 */
export const exportPlotAttributeWorkbook = (
  taskCode: string,
  payload: PlotAttributeWorkbookExportPayload,
): Promise<Blob> => request.post(
  `/v1/plot-attribute-workbooks/tasks/${encodeURIComponent(taskCode)}/export.xlsx`,
  payload,
  { responseType: 'blob', timeout: 60_000 },
)

/**
 * 服务端严格解析并原子导入地块属性 XLSX。
 * @param taskCode 作业任务编号。
 * @param file 原始工作簿实体。
 * @param operatorCode 稳定项目用户编码。
 * @param comment 影像、外业或调查表证据说明。
 * @returns 导入批次和实体校验值。
 */
export const importPlotAttributeWorkbook = (
  taskCode: string,
  file: File,
  operatorCode: string,
  comment: string,
): Promise<PlotAttributeImportBatch> => {
  const formData = new FormData()
  formData.append('file', file)
  formData.append('operator_code', operatorCode)
  formData.append('comment', comment)
  return request.post(
    `/v1/plot-attribute-workbooks/tasks/${encodeURIComponent(taskCode)}/import`,
    formData,
    {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 120_000,
    },
  )
}

/**
 * 查询任务最近的属性工作簿导入证据。
 * @param taskCode 作业任务编号。
 * @returns 最近 20 个实体导入批次。
 */
export const getPlotAttributeImportBatches = (
  taskCode: string,
): Promise<PlotAttributeImportBatchList> => request.get(
  `/v1/plot-attribute-workbooks/tasks/${encodeURIComponent(taskCode)}/imports`,
)

File: backend/app/api/v1/endpoints/plot_attribute_workbook.py
"""任务地块属性 Excel 导入导出端点。"""

from typing import Annotated

from fastapi import APIRouter, File, Form, Response, UploadFile

from app.api.deps import DatabaseSession
from app.schemas.plot_attribute_workbook import (
    PlotAttributeImportBatchListResponse,
    PlotAttributeImportBatchResponse,
    PlotAttributeWorkbookExportRequest,
    PlotAttributeWorkbookImportMetadata,
)
from app.services.plot_attribute_workbook_parser import XLSX_MEDIA_TYPE
from app.services.plot_attribute_workbook_service import (
    PlotAttributeWorkbookService,
)

router = APIRouter(
    prefix="/api/v1/plot-attribute-workbooks",
    tags=["地块属性 Excel"],
)
service = PlotAttributeWorkbookService()


@router.post(
    "/tasks/{task_code}/export.xlsx",
    response_class=Response,
    responses={200: {"content": {XLSX_MEDIA_TYPE: {}}}},
)
async def export_plot_attribute_workbook(
    task_code: str,
    request: PlotAttributeWorkbookExportRequest,
    db: DatabaseSession,
) -> Response:
    """导出任务全部或显式选择图斑的当前属性工作簿。

    Args:
        task_code: 作业任务编号。
        request: 操作人编码和可选显式图斑编号。
        db: FastAPI 注入的异步数据库会话。

    Returns:
        Response: 包含当前版本和逐行属性的 XLSX 实体。
    """
    workbook = await service.export_workbook(db, task_code, request)
    return Response(
        content=workbook.content,
        media_type=XLSX_MEDIA_TYPE,
        headers={
            "Content-Disposition": f'attachment; filename="{workbook.filename}"',
            "X-Workbook-Row-Count": str(workbook.row_count),
        },
    )


@router.post(
    "/tasks/{task_code}/import",
    response_model=PlotAttributeImportBatchResponse,
)
async def import_plot_attribute_workbook(
    task_code: str,
    file: Annotated[UploadFile, File(description="地块属性 XLSX 实体文件")],
    operator_code: Annotated[str, Form(min_length=1, max_length=50)],
    comment: Annotated[str, Form(min_length=2, max_length=500)],
    db: DatabaseSession,
) -> PlotAttributeImportBatchResponse:
    """严格校验并原子导入逐图斑属性工作簿。

    Args:
        task_code: 作业任务编号。
        file: 服务端解析的原始 XLSX 实体。
        operator_code: 项目用户稳定编码。
        comment: 影像、外业或调查表证据说明。
        db: FastAPI 注入的异步数据库会话。

    Returns:
        PlotAttributeImportBatchResponse: 批次、变更数量和实体校验值。
    """
    filename = file.filename or "plot_attributes.xlsx"
    content = await file.read()
    await file.close()
    metadata = PlotAttributeWorkbookImportMetadata(
        operator_code=operator_code,
        comment=comment,
    )
    return await service.import_workbook(
        db,
        task_code,
        metadata,
        filename,
        content,
    )


@router.get(
    "/tasks/{task_code}/imports",
    response_model=PlotAttributeImportBatchListResponse,
)
async def list_plot_attribute_import_batches(
    task_code: str,
    db: DatabaseSession,
) -> PlotAttributeImportBatchListResponse:
    """查询任务最近的地块属性 Excel 导入证据。

    Args:
        task_code: 作业任务编号。
        db: FastAPI 注入的异步数据库会话。

    Returns:
        PlotAttributeImportBatchListResponse: 最近 20 个实体导入批次。
    """
    return await service.list_import_batches(db, task_code)
